### PR TITLE
Request envelope requires "data"

### PR DESCRIPTION
**Pull request by jamesaimonetti**
Had an issue in training where user sent POST -d '{"real":"stuff"}' and was getting schema validation errors saying "real" is required. That is a confusing because the real error is that the user didn't create the request envelope properly. This PR checks that and returns a more appropriate error payload.

### CODE REVIEW

File: applications/crossbar/src/api_util.erl [hunk 1]
```diff
@@ -245,14 +226,25 @@ handle_url_encoded_body(Context, Req, QS, ReqBody, JObj) ->
...
+set_request_data_in_context(Context, Req, JObj, QS) ->
+    case is_valid_request_envelope(JObj) of
+        'false' ->
+            lager:info("failed to find 'data' in envelope, invalid request"),
+            halt_on_invalid_envelope(Req, Context);
+        'true' ->
             Setters = [{fun cb_context:set_req_json/2, JObj}
-                       ,{fun cb_context:set_req_data/2, wh_json:get_value(<<"data">>, JObj, wh_json:new())}
-                       ,{fun cb_context:set_query_string/2, QS}
+                      ,{fun cb_context:set_req_data/2, wh_json:get_value(<<"data">>, JObj)}
```
Comment (fenollp) [inline applications/crossbar/src/api_util.erl]: Don't cb_ modules depend on req_data being a JObj? (instead of undefined)

Comment (jamesaimonetti) [inline applications/crossbar/src/api_util.erl]: If we get here, we've verified `data` exists in the request data.

**Comment by lazedo** [comment]
@jamesaimonetti can you check if OPTIONS verb does not require `data`. it seems we missed this https://2600hz.atlassian.net/browse/KAZOO-4626

**Comment by fenollp** [comment]
This breaks backwards compatibility of KNM @jamesaimonetti @k-anderson

**Comment by jamesaimonetti** [comment]
@fenollp how so? PUT/POST has always required a "data" object, even an empty one.

**Comment by jamesaimonetti** [comment]
Perhaps we could amend it so that if there is no request body, we continue processing with `wh_json:new()` as req_data?

**Comment by fenollp** [comment]
Wrong. Checkout 3.22 and PUT a number to a `wnm_allow_additions` `true` account.

**Comment by fenollp** [comment]
I believe this would fix it, yes :)

**Comment by jamesaimonetti** [comment]
@fenollp see if #1773 addresses the issue

**Pull request merged**